### PR TITLE
Release 0.6.0 with canonical OCF entity typing

### PR DESCRIPTION
**Pull request by HardlyDifficult**
## Summary

- release the canonical OCF typing boundary as `@open-captable-protocol/canton@0.6.0`
- require an exact readonly `object_type` discriminator on every canonical top-level OCF model
- validate raw OCF objects against the declared schema before ledger canonicalization
- keep current-schema PlanSecurity wrappers only at raw ingestion and normalize them to EquityCompensation
- correlate all 48 entity keys with native data, generated DAML payloads, template identity, batch tags, and positive public read coverage
- validate ledger template identity and generated payloads for every `OcpClient` namespace and `getByObjectType`
- preserve canonical StockIssuance output by omitting empty optional arrays

## Breaking changes

- typed OCF inputs now require their canonical `object_type`
- PlanSecurity entity keys are no longer accepted by canonical batch/read APIs
- schema-invalid raw input is rejected instead of being repaired before validation
- high-level reads reject wrong-template and malformed generated payloads

## Release behavior

- the next main-branch publish uses the intentional manifest boundary `0.6.0` instead of auto-incrementing to `0.6.1`
- later publishes resume normal patch increments from `0.6.0`
- verified artifact: `@open-captable-protocol/canton@0.6.0`, SHA-256 `f4dcee49400867f4a305a7377ed59be3b750515371279e500e86dbfb266b0265`

## Direct-consumer validation

All organization-wide direct consumers have migration PRs built from their current main branches and validated against the exact packed artifact:

- [Fairmint/canton-fairmint-sdk#177](https://github.com/Fairmint/canton-fairmint-sdk/pull/177) — peer range `>=0.6.0 <0.7.0`, exact dev artifact validation, 459 tests
- [Fairmint/open-captable-protocol-daml#286](https://github.com/Fairmint/open-captable-protocol-daml/pull/286) — replay path migrated to canonical operation construction; schema, DAML, upgrade, codegen, replay, and scenario checks pass
- [Fairmint/canton#620](https://github.com/Fairmint/canton/pull/620) — exact registry pins, runtime capability guards, build and 253 tests pass
- [Fairmint/canton-explorer#162](https://github.com/Fairmint/canton-explorer/pull/162) — clean npm 11 install, clean full dependency tree, production build, and fail-closed retired-provider migration coverage

Downstream CI installation failures are expected until `0.6.0` is published. Canton and Explorer additionally wait for the resulting Fairmint SDK `0.0.126` publish.

## Validation

- full unit suite: 59 suites / 2,751 tests
- fixture-backed positive reads for all 48 canonical entity types through both public read APIs
- strict-ingestion coverage for all seven current-schema PlanSecurity wrappers
- canonical stakeholder-event fixtures and explicit rejection coverage for non-schema names
- build, declaration tests, lint, format, and diff checks
- package artifact verified as `@open-captable-protocol/canton@0.6.0`
- full OCP quickstart integration passes on the current head

---

> [!NOTE]
> **High Risk**
> Breaking public typing and batch APIs across all cap-table read/write paths, with stricter ledger template and payload validation that can reject previously tolerated data.
> 
> **Overview**
> Ships **@open-captable-protocol/canton@0.6.0** and tightens the public OCF boundary so reads and batch writes align with generated DAML types instead of ad hoc casts and duplicate maps.
> 
> **Canonical models and registry:** Every ledger-backed entity now carries a required **`object_type`** on native read output. **`ENTITY_REGISTRY`** is the single source for object types, **template IDs**, batch tags, and **`getByObjectType`** routing (PlanSecurity **entity keys are removed** from batch/read dispatch; schema-level PlanSecurity wrappers normalize to EquityCompensation elsewhere). Stock plan create-args use **`plan_data`**.
> 
> **Reads (`OcpClient`):** Per-entity `get*AsOcf` wiring is replaced by **`getEntityAsOcf`** plus a typed **`object_type` → reader** map. Ledger reads check **expected template ID** and **decode** payloads with generated DAML decoders before **`convertToOcf`**.
> 
> **Writes (`CapTableBatch`):** Create/edit/delete paths build union variants via **`buildOcfCreateData` / `buildOcfEditData` / `buildOcfDeleteData`** (decoder-validated) and support **`createOperation` / `editOperation` / `deleteOperation`** for pre-correlated ops.
> 
> **Release tooling:** **`prepare-release`** publishes an intentional **`package.json`** version when it is ahead of NPM and unused (e.g. **0.6.0**), then resumes patch bumps. ESLint includes **`tsconfig.declaration-tests.json`**.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 87cd44cb3a5abb7fa0f1d22f4943e02dc7522799. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: src/utils/planSecurityAliases.ts [hunk 1]
```diff
@@ -858,9 +896,16 @@ export function deepNormalizeNumericStrings<T>(value: T): T {
...
  */
-export function normalizeOcfData<T extends Record<string, unknown>>(data: T): T {
+export function normalizeOcfData(data: object): Record<string, unknown> {
+  if (
+    Array.isArray(data) ||
+    (Object.getPrototypeOf(data) !== Object.prototype && Object.getPrototypeOf(data) !== null)
+  ) {
+    throw new Error('Invalid OCF data: expected a plain object');
+  }
+  const input = data as Record<string, unknown>;
```
[the remark this answers is not in the document]
Comment (HardlyDifficult) [reply src/utils/planSecurityAliases.ts]: Fixed in 365067b. `normalizeOcfData` now accepts `unknown`, rejects null/non-objects with the intended validation error, and reads the prototype once. The regression table now includes null.

File: src/utils/typeGuards.ts [hunk 2]
```diff
@@ -81,6 +82,20 @@ export function isMonetary(value: unknown): value is { amount: string; currency:
...
+  value: unknown,
+  objectType: T['object_type']
+): value is T {
+  if (!isObject(value) || Array.isArray(value) || value.object_type !== objectType) return false;
+
+  try {
+    return getOcfSchema(objectType).safeParse(value).success;
+  } catch {
+    return false;
+  }
```
[the remark this answers is not in the document]
Comment (HardlyDifficult) [reply src/utils/typeGuards.ts:96]: Keeping the schema-backed behavior intentionally. This SDK targets Node/Canton, and the published package copies the OCF schemas into `dist`; a structural fallback would turn a failed strict validation into a weaker success and defeat this PR's contract. Returning false when strict validation cannot run is the safe type-guard result. Browser-only structural guards, if needed, should be a separate explicitly weaker API.

File: test/utils/ocfZodSchemas.test.ts [hunk 3]
```diff
@@ -15,6 +20,65 @@ function toRecord(value: unknown): Record<string, unknown> {
...
+    sourceObjectType: 'TX_PLAN_SECURITY_RETRACTION',
+    canonicalObjectType: 'TX_EQUITY_COMPENSATION_RETRACTION',
+    fixture: () => loadSyntheticFixture<Record<string, unknown>>('equityCompensationRetraction'),
+  },
+  {
+    sourceObjectType: 'TX_PLAN_SECURITY_TRANSFER',
+    canonicalObjectType: 'TX_EQUITY_COMPENSATION_TRANSFER',
+    fixture: () => loadSyntheticFixture<Record<string, unknown>>('equityCompensationTransfer'),
+  },
+] as const;
```
[the remark this answers is not in the document]
Comment (HardlyDifficult) [reply test/utils/ocfZodSchemas.test.ts]: Addressed in e9dc120. PlanSecurity cases are now generated from PLAN_SECURITY_OBJECT_TYPE_MAP, while the manual fixture loaders are separately keyed with a Record<PlanSecurityObjectType, loader> compile-time contract. The schema suite passes as part of the 251 focused tests and 2,534-test full run.

File: test/converters/convertibleCancellationConverters.test.ts [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        security_id: 'convertible-security-2',
+        reason_text: 'Missing amount',
+        comments: [],
+      },
+    });
+
+    await expect(
+      getConvertibleCancellationAsOcf(client, { contractId: 'convertible-cancellation-contract-2' })
+    ).rejects.toThrow(OcpValidationError);
+  });
```
[the remark this answers is not in the document]
Comment (HardlyDifficult) [reply test/converters/convertibleCancellationConverters.test.ts]: Addressed in e9dc120. The converter suite now calls damlConvertibleCancellationToNative directly with no LedgerJsonApiClient mock. Getter fetch behavior and the missing-amount boundary case now live in the DAML-to-OCF getter validation suite. The five focused suites pass 251 tests; full test:ci passes 56 suites and 2,534 tests.

File: src/functions/OpenCapTable/capTable/ocfToDaml.ts [hunk 5]
```diff
@@ -74,6 +78,15 @@ import { warrantTransferDataToDaml } from '../warrantTransfer/warrantTransferDat
  */
 export function convertToDaml(...args: OcfEntityArguments): Record<string, unknown> {
   const [type, data] = args;
+  return convertEntityToDaml(type, data);
+}
+
+/** Convert a correlated create/edit operation object to its generated DAML payload. */
+export function convertOperationToDaml(operation: OcfCreateOperation | OcfEditOperation): Record<string, unknown> {
+  return convertEntityToDaml(operation.type, operation.data);
+}
```
[the remark this answers is not in the document]
Comment (HardlyDifficult) [reply src/functions/OpenCapTable/capTable/ocfToDaml.ts:87]: Addressed in e9dc120. The cap-table barrel now re-exports convertOperationToDaml, and both the source compile contract and built-declaration contract import it from the package root and verify kind/payload correlation. npm run typecheck, build, and test:declarations all pass.

[the remark this answers is not in the document]
Comment (HardlyDifficult) [reply src/functions/OpenCapTable/capTable/ocfToDaml.ts:87]: Follow-up re-evaluation in 0cc6a29: we intentionally removed the barrel export. `convertOperationToDaml` is an internal sibling helper used to preserve correlated implementation typing, not a supported package API. The validated public builders and canonical `convertToDaml` remain exposed, and the downstream curated root API in #412 likewise excludes low-level conversion/read helpers. Typecheck, build, declarations, lint, formatting, and the 251 affected tests all pass.

File: src/utils/ocfMetadata.ts [hunk 6]
```diff
@@ -1,7 +1,7 @@
 import { Fairmint } from '@fairmint/open-captable-protocol-daml-js';
 
-/** OCF object types supported by the SDK */
-export type OcfObjectType =
+/** OCF object types covered by the legacy template-metadata registry. */
+export type OcfMetadataObjectType =
```
[the remark this answers is not in the document]
Comment (HardlyDifficult) [reply src/utils/ocfMetadata.ts:4]: Not applying this compatibility alias. The SDK redesign is intentionally breaking and the user explicitly stated that backwards compatibility is not required. OcfMetadataObjectType is the curated public name; retaining the ambiguous OcfObjectType alias would preserve the legacy surface this PR is deliberately removing.

File: src/functions/OpenCapTable/stockPlan/getStockPlanAsOcf.ts [hunk 7]
```diff
@@ -114,21 +101,16 @@ export async function getStockPlanAsOcf(
...
 
-  const native = damlStockPlanDataToNative(
-    createArgument.plan_data as Fairmint.OpenCapTable.OCF.StockPlan.StockPlanOcfData
-  );
-
-  const ocf: OcfStockPlanOutput = {
-    object_type: 'STOCK_PLAN',
-    ...native,
-  };
+  const stockPlan = damlStockPlanDataToNative(planData as Fairmint.OpenCapTable.OCF.StockPlan.StockPlanOcfData);
```
[the remark this answers is not in the document]
Comment (HardlyDifficult) [reply src/functions/OpenCapTable/stockPlan/getStockPlanAsOcf.ts]: Addressed in be84650. `getStockPlanAsOcf` now decodes `plan_data` with the generated `StockPlanOcfData.decoder` before canonicalization and maps decoder failures to `OcpParseError` / `SCHEMA_MISMATCH`; the broad cast is gone. Regression coverage includes `{}`, wrong-field objects, missing required fields, primitive/container failures, valid and zero-value payloads, plus direct-converter non-object safety. Full validation passes: 56 suites / 2,542 tests, typecheck, build, declarations, lint, formatting, and diff check.

File: src/functions/OpenCapTable/convertibleCancellation/getConvertibleCancellationAsOcf.ts [hunk 8]
```diff
@@ -1,7 +1,9 @@
 import type { LedgerJsonApiClient } from '@fairmint/canton-node-sdk';
 import { type Fairmint } from '@fairmint/open-captable-protocol-daml-js';
 import type { GetByContractIdParams } from '../../../types/common';
```
[the remark this answers is not in the document]
Comment (HardlyDifficult) [reply src/functions/OpenCapTable/convertibleCancellation/getConvertibleCancellationAsOcf.ts:3]: Addressed in e08b4ca: `Fairmint` is now a runtime import so this reader can use the generated `ConvertibleCancellation.templateId`. Clean build, declarations, full coverage (57 suites / 2,687 tests), lint, formatting, and diff checks pass.

File: src/functions/OpenCapTable/convertibleCancellation/getConvertibleCancellationAsOcf.ts [hunk 9]
```diff
@@ -47,14 +41,14 @@ export async function getConvertibleCancellationAsOcf(
   const contract = createArgument as ConvertibleCancellationCreateArgument;
   const data = contract.cancellation_data;
```
[the remark this answers is not in the document]
Comment (HardlyDifficult) [reply src/functions/OpenCapTable/convertibleCancellation/getConvertibleCancellationAsOcf.ts:43]: Addressed in e08b4ca: `readSingleContract` now receives the generated ConvertibleCancellation `expectedTemplateId`, and a wrong-template regression proves fail-closed `SCHEMA_MISMATCH` behavior. Full local coverage passes: 57 suites / 2,687 tests.

File: src/functions/OpenCapTable/capTable/CapTableBatch.ts [hunk 10]
```diff
@@ -67,6 +71,108 @@ export interface BatchItemMeta {
...
+    throw new OcpValidationError(
+      `batch.${operation}.${entityType}`,
+      `Converter output does not match the generated DAML ${operation} variant: ${message}`,
+      {
+        code: OcpErrorCodes.INVALID_FORMAT,
+        receivedValue: input,
+      }
+    );
+  }
+}
```
[the remark this answers is not in the document]
Comment (HardlyDifficult) [reply src/functions/OpenCapTable/capTable/CapTableBatch.ts:93]: Addressed in b6103f45: generated-operation decode errors no longer retain the raw converted payload. The error carries only operation/entityType context, and a forced-failure regression asserts `receivedValue` is undefined and serialized error data cannot contain the stakeholder ID. Clean build, declarations, lint/format, focused 145 tests, and full coverage (57 suites / 2,688 tests) pass.

File: src/OcpClient.ts [hunk 11]
```diff
@@ -590,285 +544,84 @@ export class OcpClient {
...
-        },
-      },
-      stakeholderStatusChangeEvent: {
-        get: async (params) => {
-          const r = await getStakeholderStatusChangeEventAsOcf(client, params);
-          return toContractResult(r.event, r.contractId);
-        },
-      },
+      stakeholderRelationshipChangeEvent: genericEntity('stakeholderRelationshipChangeEvent'),
+      stakeholderStatusChangeEvent: genericEntity('stakeholderStatusChangeEvent'),
```
[the remark this answers is not in the document]
Comment (HardlyDifficult) [reply src/OcpClient.ts:624]: Addressed in 87cd44cb: the shared DAML-to-OCF relationship-event converter now rejects payloads where both relationship_started and relationship_ended are absent. The regression exercises the OcpClient namespace and asserts controlled INVALID_FORMAT behavior. Full validation passes: build, declarations, lint, formatting, 59 suites, and 2,751 tests.
